File: partyapp/src/main/java/com/example/partyapp/Adapters/ServersAdapter.java
package com.example.partyapp.Adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.TextView;

import com.example.partyapp.Models.Server;
import com.example.partyapp.R;

import java.util.ArrayList;

public class ServersAdapter extends BaseAdapter {

    ArrayList<Server> servers;
    Context context;

    public ServersAdapter(ArrayList<Server> serverArrayList, Context context) {
        servers = serverArrayList;
        this.context = context;
    }
    @Override
    public int getCount() {
        return servers.size();
    }

    @Override
    public Object getItem(int position) {
        return null;
    }

    @Override
    public long getItemId(int position) {
        return 0;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        LayoutInflater inflater = LayoutInflater.from(context);
        View view = inflater.inflate(R.layout.list_item_servers, parent, false);
        TextView nameTextView = (TextView) view.findViewById(R.id.name);
        TextView distanceTextView  = (TextView) view.findViewById(R.id.distance);
        Server server = servers.get(position);

        nameTextView.setText(server.name);
        distanceTextView.setText(String.valueOf(server.distance) + " km");

        return view;
    }
}

File: app/src/main/java/com/tesonet/testio/ui/adapters/ServersAdapter.kt
package com.tesonet.testio.ui.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.tesonet.testio.R
import com.tesonet.testio.databinding.ServerListItemBinding
import com.tesonet.testio.services.data.server.Server
import com.tesonet.testio.ui.adapters.ServersAdapter.ViewHolder
import com.tesonet.testio.utils.addKilometers

class ServersAdapter(private val context: Context) : RecyclerView.Adapter<ViewHolder>() {

    private val servers = ArrayList<Server>()
    var selectedServer: ((Server) -> Unit) = {}

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(ServerListItemBinding.inflate(LayoutInflater.from(context), parent, false).root)
    }

    override fun getItemCount(): Int {
        return servers.size
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val server = servers[position]
        holder.run {
            serverName.text = server.name
            serverDistance.text = server.distance.addKilometers()

            itemView.setOnClickListener {
                selectedServer(server)
            }
        }
    }

    fun setItem(servers: List<Server>?) {
        this.servers.clear()
        servers?.let {
            this.servers.addAll(it)
        }

        notifyDataSetChanged()
    }

    class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {

        val serverName: TextView = view.findViewById(R.id.textViewServerName)
        val serverDistance: TextView = view.findViewById(R.id.textViewServerDistance)
    }
}